### PR TITLE
Make GT side clamp in select_candidates_in_gts monotonic

### PR DESCRIPTION
**Pull request by ShuaiLYU**
The threshold used stride[0] (8px) while the replacement used stride_val (stride[1], 16px), so a side of 7.9px was raised to 16px while 8.1px was left alone. Sides in [8, 16) therefore got a smaller candidate pool than sides below 8 -- averaged over random box positions on a 640px grid, a 7.9px square drew 5.28 candidate anchors and an 8.0px square only 1.32.

Comparing against stride_val makes the operation a plain floor to 16px, which is monotonic in box size. Applied to the rotated assigner too.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Updated candidate selection in the standard and rotated assigners to clamp tiny ground-truth box sides using `self.stride_val`, preserving a monotonic candidate pool as box size increases.

### 📊 Key Changes
- Changed the ground-truth side threshold in both `select_candidates_in_gts` implementations from `self.stride[0]` to `self.stride_val`.
- Applied the same clamp behavior to axis-aligned and rotated candidate assignment.
- Retained replacement of undersized sides with `self.stride_val` while respecting the ground-truth mask and tensor device/dtype.

### 🎯 Purpose & Impact
- Prevents box sides in the range below `self.stride_val` from receiving inconsistent candidate pools due to the previous threshold, improving monotonic candidate selection for small objects.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShuaiLYU, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

LGTM. The change consistently aligns the clamp threshold with the replacement value in both axis-aligned and rotated assigners, removing the candidate-pool discontinuity for sides between the smallest and second strides. No concrete regressions found in the reviewed scope.

**Comment by ShuaiLYU** [comment]
## Effect on anchor assignment

`select_candidates_in_gts` compared GT sides against `stride[0]` (8px) but replaced them with `stride_val` (`stride[1]`, 16px). A 7.9px side was raised to 16px while an 8.1px side was left untouched, so the candidate pool *shrank* as the box grew. Comparing against `stride_val` on both sides makes it a plain floor to 16px.

Mean candidate anchors per GT box, averaged over 2000 random box centers on a 640px image (strides [8, 16, 32], seed 0). Checkout implements: **after (fixed)**.

### Square boxes

| GT w x h | before | after | ratio | before P3/P4/P5 | after P3/P4/P5 |
| --- | --- | --- | --- | --- | --- |
| 4x4 | 5.24 | 5.24 | same | 4.00 / 1.00 / 0.24 | 4.00 / 1.00 / 0.24 |
| 6x6 | 5.24 | 5.24 | same | 4.00 / 1.00 / 0.24 | 4.00 / 1.00 / 0.24 |
| 7.9x7.9 | 5.24 | 5.24 | same | 4.00 / 1.00 / 0.24 | 4.00 / 1.00 / 0.24 |
| 8x8 | 1.30 | 5.24 | **4.02x** | 1.00 / 0.24 / 0.06 | 4.00 / 1.00 / 0.24 |
| 8.1x8.1 | 1.33 | 5.24 | **3.93x** | 1.03 / 0.25 / 0.06 | 4.00 / 1.00 / 0.24 |
| 10x10 | 2.04 | 5.24 | **2.56x** | 1.58 / 0.37 / 0.09 | 4.00 / 1.00 / 0.24 |
| 12x12 | 2.94 | 5.24 | **1.78x** | 2.26 / 0.54 / 0.14 | 4.00 / 1.00 / 0.24 |
| 14x14 | 4.02 | 5.24 | **1.30x** | 3.09 / 0.74 / 0.19 | 4.00 / 1.00 / 0.24 |
| 15.9x15.9 | 5.17 | 5.24 | **1.01x** | 3.95 / 0.98 / 0.23 | 4.00 / 1.00 / 0.24 |
| 16x16 | 5.24 | 5.24 | same | 4.00 / 1.00 / 0.24 | 4.00 / 1.00 / 0.24 |
| 16.1x16.1 | 5.30 | 5.30 | same | 4.05 / 1.02 / 0.24 | 4.05 / 1.02 / 0.24 |
| 20x20 | 8.20 | 8.20 | same | 6.22 / 1.61 / 0.37 | 6.22 / 1.61 / 0.37 |
| 24x24 | 11.82 | 11.82 | same | 9.00 / 2.27 / 0.55 | 9.00 / 2.27 / 0.55 |
| 32x32 | 21.00 | 21.00 | same | 16.00 / 4.00 / 1.00 | 16.00 / 4.00 / 1.00 |
| 48x48 | 47.29 | 47.29 | same | 36.00 / 9.00 / 2.29 | 36.00 / 9.00 / 2.29 |
| 64x64 | 84.00 | 84.00 | same | 64.00 / 16.00 / 4.00 | 64.00 / 16.00 / 4.00 |

### Slivers (the clamp is per-axis: only the sub-threshold side is raised)

| GT w x h | before | after | ratio | before P3/P4/P5 | after P3/P4/P5 |
| --- | --- | --- | --- | --- | --- |
| 4x32 | 10.48 | 10.48 | same | 8.00 / 2.00 / 0.48 | 8.00 / 2.00 / 0.48 |
| 7.9x32 | 10.48 | 10.48 | same | 8.00 / 2.00 / 0.48 | 8.00 / 2.00 / 0.48 |
| 8x32 | 5.21 | 10.48 | **2.01x** | 4.00 / 0.96 / 0.24 | 8.00 / 2.00 / 0.48 |
| 12x32 | 7.88 | 10.48 | **1.33x** | 6.06 / 1.45 / 0.37 | 8.00 / 2.00 / 0.48 |
| 16x32 | 10.48 | 10.48 | same | 8.00 / 2.00 / 0.48 | 8.00 / 2.00 / 0.48 |
| 4x128 | 41.92 | 41.92 | same | 32.00 / 8.00 / 1.92 | 32.00 / 8.00 / 1.92 |
| 7.9x128 | 41.92 | 41.92 | same | 32.00 / 8.00 / 1.92 | 32.00 / 8.00 / 1.92 |
| 8x128 | 20.83 | 41.92 | **2.01x** | 16.00 / 3.86 / 0.97 | 32.00 / 8.00 / 1.92 |
| 12x128 | 31.50 | 41.92 | **1.33x** | 24.22 / 5.79 / 1.49 | 32.00 / 8.00 / 1.92 |
| 16x128 | 41.92 | 41.92 | same | 32.00 / 8.00 / 1.92 | 32.00 / 8.00 / 1.92 |

Non-monotonic steps over a 4->40px sweep in 0.1px increments: **before 1, after 0**.

Boxes at 16.1px and above are unaffected, so the change is contained to sides in `[8, 16)`. The sliver table shows the clamp is per-axis: `4x32` and `16x32` are identical before/after, only the sub-threshold width moves.

<details>
<summary>Reproduction script</summary>

Save as `runs/tests/test_gts_clamp_monotonic.py` and run `python runs/tests/test_gts_clamp_monotonic.py`. The geometry test is the real in-tree `select_candidates_in_gts`, invoked through an assigner with `stride=[0, 0, 0]` so its own clamp is a no-op; only the clamp itself is expressed twice. The self-check line reports which variant the checkout implements.

```python
"""Compare candidate-anchor counts in TaskAlignedAssigner.select_candidates_in_gts before/after the clamp fix.

Before: `wh < stride[0]` (8px) selected sides, but the replacement value was `stride_val` (stride[1], 16px),
so a 7.9px side was raised to 16px while an 8.1px side was left alone -- the candidate pool shrank as the
box grew. After: both sides of the comparison use `stride_val`, making the op a plain floor to 16px.

Only the clamp is re-expressed here; the geometry test itself is the real in-tree function, invoked with the
clamp disabled. A self-check confirms which variant the current checkout actually implements.

Usage: python runs/tests/test_gts_clamp_monotonic.py  ->  writes runs/tests/gts_clamp_table.md
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))  # import the in-tree ultralytics, not the installed one

import torch

from ultralytics.utils.tal import TaskAlignedAssigner, make_anchors

IMGSZ, STRIDES, N, SEED = 640, [8, 16, 32], 2000, 0
THRESH_BEFORE, FILL = STRIDES[0], STRIDES[1]  # 8px threshold, 16px replacement

torch.manual_seed(SEED)
feats = [(IMGSZ // s, IMGSZ // s) for s in STRIDES]
anchors, stride_tensor = make_anchors(
    torch.tensor(feats, dtype=torch.float32), torch.tensor(STRIDES, dtype=torch.float32)
)
anchors = anchors * stride_tensor  # anchor centers in image pixels
lvl = stride_tensor.squeeze(-1)  # per-anchor stride, for the per-level breakdown
centers = 160 + 320 * torch.rand(N, 2)  # random centers, well inside the image

asn = TaskAlignedAssigner(stride=STRIDES)
asn_raw = TaskAlignedAssigner(stride=[0, 0, 0])  # threshold 0 -> clamp is a no-op, pure geometry test


def clamp(wh, after):
    """Apply the before- or after-fix side clamp to a (N, 2) width/height tensor."""
    return wh.clamp(min=FILL) if after else torch.where(wh < THRESH_BEFORE, torch.tensor(float(FILL)), wh)


def counts(w, h, after):
    """Return (mean total, mean per-level) candidate anchor counts over N random box centers."""
    wh = clamp(torch.tensor([float(w), float(h)]).expand(N, 2), after)
    boxes = torch.cat([centers - wh / 2, centers + wh / 2], -1).unsqueeze(1)  # (N, 1, 4)
    m = asn_raw.select_candidates_in_gts(anchors, boxes, torch.ones(N, 1, 1))[:, 0]  # (N, h*w)
    return m.sum(-1).float().mean().item(), [m[:, lvl == s].sum(-1).float().mean().item() for s in STRIDES]


def self_check():
    """Verify which variant the in-tree assigner implements, by matching it against both."""
    wh = torch.full((N, 2), 12.0)  # 12px trips the clamp only after the fix
    boxes = torch.cat([centers - wh / 2, centers + wh / 2], -1).unsqueeze(1)
    live = asn.select_candidates_in_gts(anchors, boxes, torch.ones(N, 1, 1))[:, 0].sum(-1).float().mean().item()
    return "after (fixed)" if abs(live - counts(12, 12, True)[0]) < 1e-4 else "before (unfixed)"


def nonmono(after):
    """Count non-monotonic steps over a fine size sweep."""
    prev, bad = -1.0, 0
    for i in range(361):
        cur = counts(4 + 0.1 * i, 4 + 0.1 * i, after)[0]
        bad += cur < prev - 1e-6
        prev = cur
    return bad


def row(w, h):
    """Format one markdown table row comparing before/after for a w*h GT box."""
    tb, pb = counts(w, h, False)
    ta, pa = counts(w, h, True)
    d = "same" if abs(ta - tb) < 1e-4 else f"**{ta / tb:.2f}x**"
    return (
        f"| {w:g}x{h:g} | {tb:.2f} | {ta:.2f} | {d} | "
        f"{pb[0]:.2f} / {pb[1]:.2f} / {pb[2]:.2f} | {pa[0]:.2f} / {pa[1]:.2f} / {pa[2]:.2f} |"
    )


out = [
    f"Mean candidate anchors per GT box, averaged over {N} random box centers on a {IMGSZ}px image "
    f"(strides {STRIDES}, seed {SEED}). Checkout implements: **{self_check()}**.",
    "",
    "### Square boxes",
    "",
    "| GT w x h | before | after | ratio | before P3/P4/P5 | after P3/P4/P5 |",
    "| --- | --- | --- | --- | --- | --- |",
    *(row(s, s) for s in (4, 6, 7.9, 8, 8.1, 10, 12, 14, 15.9, 16, 16.1, 20, 24, 32, 48, 64)),
    "",
    "### Slivers (the clamp is per-axis: only the sub-threshold side is raised)",
    "",
    "| GT w x h | before | after | ratio | before P3/P4/P5 | after P3/P4/P5 |",
    "| --- | --- | --- | --- | --- | --- |",
    *(row(w, h) for h in (32, 128) for w in (4, 7.9, 8, 12, 16)),
    "",
    f"Non-monotonic steps over a 4->40px sweep in 0.1px increments: "
    f"**before {nonmono(False)}, after {nonmono(True)}**.",
]
text = "\n".join(out)
print(text)
Path(__file__).with_name("gts_clamp_table.md").write_text(text + "\n")
```

</details>
